File: go/lesson-09-testing/exercises/string_utils.go
package stringutils

// Reverse returns the string with its characters reversed.
// Reverse("hello") → "olleh"
// Reverse("Go") → "oG"
func Reverse(s string) string {
	// TODO: implement
	// Hint: convert to []rune to handle Unicode correctly
	return ""
}

// IsPalindrome reports whether s reads the same forwards and backwards.
// Case-insensitive. Ignores non-letter characters.
// IsPalindrome("racecar") → true
// IsPalindrome("A man a plan a canal Panama") → true
// IsPalindrome("hello") → false
func IsPalindrome(s string) bool {
	// TODO: implement
	// Hint: use unicode.IsLetter and unicode.ToLower from the "unicode" package
	return false
}

// CountVowels returns the number of vowels (a, e, i, o, u) in s.
// Case-insensitive.
// CountVowels("hello") → 2
// CountVowels("rhythm") → 0
func CountVowels(s string) int {
	// TODO: implement
	return 0
}

// TitleCase converts a string to title case (first letter of each word uppercase).
// TitleCase("hello world") → "Hello World"
// TitleCase("the quick brown fox") → "The Quick Brown Fox"
func TitleCase(s string) string {
	// TODO: implement
	// Hint: strings.Fields, strings.ToUpper, strings.Join
	return ""
}

// Truncate shortens s to at most maxLen characters.
// If s is longer, append "..." (the "..." counts toward maxLen).
// Truncate("hello world", 8) → "hello..."
// Truncate("hi", 8) → "hi"
func Truncate(s string, maxLen int) string {
	// TODO: implement
	return ""
}

File: nodejs/lesson-09-testing/exercises/stringUtils.js
'use strict';

// Reverse a string. Handles Unicode correctly.
// reverse('hello') → 'olleh'
// reverse('Go 🚀') → '🚀 oG'
function reverse(str) {
  // TODO: spread into array, reverse, join
  return '';
}

// Return true if `str` is a palindrome (case-insensitive, letters only).
// isPalindrome('racecar') → true
// isPalindrome('A man a plan a canal Panama') → true
function isPalindrome(str) {
  // TODO: normalize (lower, remove non-letters), compare with its reverse
  return false;
}

// Count vowels (a, e, i, o, u) case-insensitively.
// countVowels('Hello World') → 3
function countVowels(str) {
  // TODO: filter/reduce
  return 0;
}

// Convert to title case.
// titleCase('the quick brown fox') → 'The Quick Brown Fox'
function titleCase(str) {
  // TODO: split on whitespace, capitalize first letter of each word, join
  return '';
}

// Truncate to maxLen chars. If truncated, end with '...' (counts toward maxLen).
// truncate('hello world', 8) → 'hello...'
// truncate('hi', 8) → 'hi'
function truncate(str, maxLen) {
  // TODO: implement
  return '';
}

module.exports = { reverse, isPalindrome, countVowels, titleCase, truncate };
